pkg/agent/client.go

```diff
@@ -538,10 +538,19 @@ func (a *Client) remoteToProxy(connID int64, eConn *endpointConn) {
 	}
 
 	for {
+		select {
+		case <-a.stopCh:
+			return
+		default:
+		}
+		timeout := time.Now().Add(15 * time.Second)
+		eConn.conn.SetReadDeadline(timeout)
```

Comment:
I think what we want to do here is close the connection if stopch is closed. Something like this?

```go
go func() {
  <-a.stopCh
  eConn.conn.Close()
}
```

Reply:
This function has a defer cleanup() call on line 533. When we return it should get called. That cleanup makes a protected call to https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/c1a1e092d30133901cddf3ada0b41678f95ca0e7/pkg/agent/client.go#L386. That will then close() the connection.

Reply:
Is it a problem to call `close()` multiple times? I wasn't suggesting that we were leaving the connection open, but rather calling close should be a way to quickly react to the stop channel, rather than polling for it every 15 seconds, which is what I understood your change to be doing?

Reply:
I don't believe close in golang is re-entrant. In fact I think the reason we had to protect the clean up with call once in that code path is specifically because call causes a panic if called twice.

Reply:
maybe just call `eConn.cleanup()` then?

Reply:
We called `defer eConn.cleanup()` on line 533. So returning should mean it gets called.

Reply:
tallclair is going to try the separate goroutine suggestion.